### PR TITLE
SONAR-8622 Lazily generate metadata for input files

### PR DESCRIPTION
**Pull request by dbmeneses**
Some unit tests and javadoc missing (will do it once this is validated with the other tickets).
Most files changed are tests because of DefaultInputFile.
I won't merge before Monday.

### CODE REVIEW

File: plugins/sonar-xoo-plugin/src/test/java/org/sonar/xoo/coverage/ItCoverageSensorTest.java [hunk 1]
```diff
@@ -60,7 +61,7 @@ public void testDescriptor() {
 
   @Test
   public void testNoExecutionIfNoCoverageFile() {
-    DefaultInputFile inputFile = new DefaultInputFile("foo", "src/foo.xoo").setLanguage("xoo");
+    InputFile inputFile = new TestInputFileBuilder("foo", "src/foo.xoo").setLanguage("xoo").build();
```
Comment (henryju) [inline plugins/sonar-xoo-plugin/src/test/java/org/sonar/xoo/coverage/ItCoverageSensorTest.java]: What about having a static DefaultInputFile.newBuilder("foo", "src/foo.xoo")

Comment (dbmeneses) [inline plugins/sonar-xoo-plugin/src/test/java/org/sonar/xoo/coverage/ItCoverageSensorTest.java]: It should only be exposed in tests IMO

Comment (dbmeneses) [inline plugins/sonar-xoo-plugin/src/test/java/org/sonar/xoo/coverage/ItCoverageSensorTest.java]: And this decouples InputFile from DefaultInputFile in the tests. Note that it involves IndexedFile too.

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/IndexedFile.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+   * @since 5.1
+   */
+  @Override
+  Path path();
+
+  /**
+   * Language, for example "java" or "php". Can be null if indexation of all files is enabled and no language claims to support the file.
+   */
+  @CheckForNull
+  String language();
```
Comment (henryju) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/IndexedFile.java]: I'm wondering if we should not hide the concept of language in a new API...

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/DefaultModuleFileSystem.java [hunk 3]
```diff
@@ -46,22 +46,26 @@
   private List<File> testDirsOrFiles = Lists.newArrayList();
   private ComponentIndexer componentIndexer;
   private boolean initialized;
+  private Charset charset = null;
 
   public DefaultModuleFileSystem(ModuleInputFileCache moduleInputFileCache, Project project,
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/DefaultModuleFileSystem.java]: MAJOR Constructor has 8 parameters, which is greater than 7 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00107)

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/InputFile.java [hunk 4]
```diff
@@ -41,7 +41,7 @@
  *
  * @since 4.2
  */
-public interface InputFile extends InputPath {
+public interface InputFile extends IndexedFile, InputPath {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/InputFile.java]: MINOR "InputPath" is already extended by "IndexedFile"; there is no need to implement it here. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1939)

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/internal/InputFilePredicate.java [hunk 5]
```diff
@@ -21,20 +21,7 @@
...
-    this.status = status;
-  }
-
-  @Override
-  public boolean apply(InputFile f) {
-    return status == f.status();
-  }
-
+@FunctionalInterface
+public interface InputFilePredicate {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/internal/InputFilePredicate.java]: MAJOR Drop this interface in favor of "java.util.function.Predicate<InputFile>". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1711)

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/InputFileFilter.java [hunk 6]
```diff
@@ -28,6 +28,7 @@
  */
 @ScannerSide
 @ExtensionPoint
+@FunctionalInterface
 public interface InputFileFilter {
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/InputFileFilter.java]: MAJOR Drop this interface in favor of "java.util.function.Predicate<InputFile>". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1711)

File: sonar-scanner-engine/src/test/java/org/sonar/scanner/scan/filesystem/MetadataGeneratorTest.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+import org.sonar.api.batch.fs.internal.DefaultIndexedFile;
+import org.sonar.api.batch.fs.internal.DefaultInputFile;
+import org.sonar.api.batch.fs.internal.FileMetadata;
+import org.sonar.api.batch.fs.internal.Metadata;
+import org.sonar.api.batch.fs.internal.TestInputFileBuilder;
+import org.sonar.api.config.MapSettings;
+import org.sonar.api.scan.filesystem.PathResolver;
+import org.sonar.api.utils.PathUtils;
+
+public class MetadataGeneratorTest {
```
Comment (SonarTech) [inline sonar-scanner-engine/src/test/java/org/sonar/scanner/scan/filesystem/MetadataGeneratorTest.java]: BLOCKER Add some tests to this class. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2187)

File: sonar-scanner-engine/src/test/java/org/sonar/scanner/scan/filesystem/MetadataGeneratorTest.java [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
```
Comment (SonarTech) [inline sonar-scanner-engine/src/test/java/org/sonar/scanner/scan/filesystem/MetadataGeneratorTest.java]: MINOR Remove this unused "TemporaryFolder". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2924)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java [hunk 9]
```diff
@@ -146,4 +75,18 @@ private static Charset detectCharset(File file, Charset defaultCharset) {
...
   }
+
+  public Metadata readMetadata(final DefaultInputFile inputFile, Charset defaultEncoding) {
+    try {
+      Charset charset = detectCharset(inputFile.file(), defaultEncoding);
+      inputFile.setCharset(charset);
+      Metadata metadata = fileMetadata.readMetadata(inputFile.file(), charset);
+      inputFile.setStatus(statusDetection.status(inputFile.moduleKey(), inputFile.relativePath(), metadata.hash()));
+      LOG.debug("'{}' generated metadata {} with and charset '{}'",
+        inputFile.relativePath(), inputFile.type() == Type.TEST ? "as test " : "", charset);
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java]: MAJOR Invoke method(s) only conditionally. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2629)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/FileIndexer.java [hunk 10]
```diff
@@ -107,99 +98,71 @@ void index(DefaultModuleFileSystem fileSystem) {
...
+    DefaultIndexedFile indexedFile = indexedFileBuilder.create(realFile, type, fileSystem.baseDirPath());
+    if (indexedFile != null) {
+      if (exclusionFilters.accept(indexedFile, type)) {
+        InputFile inputFile = new DefaultInputFile(indexedFile, f -> metadataGenerator.readMetadata(f, fileSystem.encoding()));
+        if (accept(inputFile)) {
+          fileSystem.add(inputFile);
+        }
+        indexParentDir(fileSystem, indexedFile);
+        progress.markAsIndexed(indexedFile);
+        LOG.debug("'{}' indexed {} with language '{}'", indexedFile.relativePath(), type == Type.TEST ? "as test " : "", indexedFile.language());
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/FileIndexer.java]: MAJOR Invoke method(s) only conditionally. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2629)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/FileIndexer.java [hunk 11]
```diff
@@ -107,99 +98,71 @@ void index(DefaultModuleFileSystem fileSystem) {
...
-        String relativePath = new PathResolver().relativePath(fs.baseDir(), parentDir);
-        if (relativePath != null) {
-          DefaultInputDir inputDir = new DefaultInputDir(fs.moduleKey(), relativePath);
-          fs.add(inputDir);
-        }
-      }
-      return null;
-    }));
-
+  private void indexParentDir(DefaultModuleFileSystem fileSystem, IndexedFile indexedFile) {
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/FileIndexer.java]: MAJOR Make "indexParentDir" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/SameInputFilePredicate.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    if (StringUtils.isEmpty(previousHash)) {
+      // ADDED
+      return true;
+    }
+
+    // this will trigger computation of metadata
+    String hash = ((DefaultInputFile) inputFile).hash();
+    if (StringUtils.equals(hash, previousHash)) {
+      // SAME
+      LOG.debug("'{}' filtering unmodified file", inputFile.relativePath());
```
Comment (SonarTech) [inline sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/SameInputFilePredicate.java]: MAJOR Invoke method(s) only conditionally. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2629)

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/internal/PathPatternPredicate.java [hunk 13]
```diff
@@ -19,6 +19,7 @@
  */
 package org.sonar.api.batch.fs.internal;
 
+import org.sonar.api.batch.fs.IndexedFile;
 import org.sonar.api.batch.fs.InputFile;
```
Comment (SonarTech) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/internal/PathPatternPredicate.java]: MINOR Remove this unused import 'org.sonar.api.batch.fs.InputFile'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 14 issues
* BLOCKER 1 blocker
* CRITICAL 2 critical
* MAJOR 8 major
* MINOR 3 minor

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [MetadataGenerator.java#L63](https://github.com/SonarSource/sonarqube/blob/fc44e0745030de6e8e982f62776246f78d5d0d3a/sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java#L63): Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1067)
1. CRITICAL [MetadataGenerator.java#L65](https://github.com/SonarSource/sonarqube/blob/fc44e0745030de6e8e982f62776246f78d5d0d3a/sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java#L65): Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1067)
1. MAJOR [MetadataGenerator.java#L57](https://github.com/SonarSource/sonarqube/blob/fc44e0745030de6e8e982f62776246f78d5d0d3a/sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/MetadataGenerator.java#L57): Reduce the number of returns of this method 6, down to the maximum allowed 5. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1142)

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/internal/DefaultInputFile.java [hunk 14]
```diff
@@ -21,274 +21,240 @@
...
-  public DefaultInputFile(String moduleKey, String relativePath) {
-    this.moduleKey = moduleKey;
-    this.relativePath = PathUtils.sanitize(relativePath);
+  public DefaultInputFile(DefaultIndexedFile indexedFile, Function<DefaultInputFile, Metadata> metadataGenerator) {
+    this.indexedFile = indexedFile;
+    this.metadataGenerator = metadataGenerator;
+    this.metadata = null;
+  }
+
+  private void checkMetadata() {
```
Comment (henryju) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/internal/DefaultInputFile.java]: Shouldn't we already prepare for multi-threading? This one would need to be synchronized.

File: sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/internal/TestInputFileBuilder.java [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+import java.nio.charset.Charset;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import javax.annotation.Nullable;
+import org.sonar.api.batch.fs.InputFile;
+import org.sonar.api.utils.PathUtils;
+
+/**
+ * @since 4.2
```
Comment (henryju) [inline sonar-plugin-api/src/main/java/org/sonar/api/batch/fs/internal/TestInputFileBuilder.java]: 4.2??

File: sonar-scanner-engine/src/test/java/org/sonar/scanner/scan/filesystem/MetadataGeneratorTest.java [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+
+  private MetadataGenerator builder;
+
+  @Before
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+    builder = new MetadataGenerator(statusDetection, metadata);
+    // TODO
+  }
+  /*
```
Comment (henryju) [inline sonar-scanner-engine/src/test/java/org/sonar/scanner/scan/filesystem/MetadataGeneratorTest.java]: What is all this commented out code?

**Pull request closed without merging**